src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,348 @@
+                    return isMemberAccessible(symbol.ContainingType, symbol.DeclaredAccessibility, within, symbol.IsStatic ? null : throughTypeOpt);
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+            }
+
+            bool isNamedTypeAccessibleWithin(INamedTypeSymbol type, ISymbol within0)
+            {
+                Debug.Assert(within0 is INamedTypeSymbol || within0 is IAssemblySymbol);
+                Debug.Assert(type != null);
+
+                if (!type.IsDefinition)
+                {
+                    foreach (var typeArg in type.TypeArguments)
+                    {
+                        if (!IsAccessibleWithin(typeArg, within0))
+                        {
+                            return false;
+                        }
+                    }
+                }
+
+                var containingType = type.ContainingType;
+                if (containingType == null)
+                {
+                    return isNonNestedTypeAccessible(type.ContainingAssembly, type.DeclaredAccessibility, within0);
+                }
+                else
+                {
+                    return isMemberAccessible(containingType, type.DeclaredAccessibility, within0, throughTypeOpt0: null);
+                }
+            }
+
+            bool isNonNestedTypeAccessible(IAssemblySymbol assembly, Accessibility declaredAccessibility, ISymbol within0)
+            {
+                Debug.Assert(within0 is INamedTypeSymbol || within0 is IAssemblySymbol);
+                Debug.Assert(assembly != null);
+
+                switch (declaredAccessibility)
+                {
+                    case Accessibility.Public:
+                        // Public symbols are always accessible from any context
+                        return true;
+
+                    case Accessibility.Private:
+                    case Accessibility.Protected:
+                    case Accessibility.ProtectedAndInternal:
+                        // Shouldn't happen except in error cases as these access levels cannot be used to declare top-level types.
+                        return false;
+
+                    case Accessibility.Internal:
+                    case Accessibility.ProtectedOrInternal:
+                        // An internal type is accessible if we're in the same assembly or we have
+                        // friend access to the assembly it was defined in.
+                        var withinAssembly = (within0 as IAssemblySymbol) ?? ((INamedTypeSymbol)within0).ContainingAssembly;
+                        return hasInternalAccessTo(withinAssembly, assembly);
+
+                    case Accessibility.NotApplicable:
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(declaredAccessibility);
+                }
+            }
+
+            bool hasInternalAccessTo(IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly)
```

Comment:
>IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly) [](start = 37, length = 57)

The names of the parameters are confusing and I would say even somewhat backwards. It looks like we are going to check if ```toAssembly``` is giving access to ```fromAssembly```, it would be more intuitive if the names were reversed, I think.

Reply:
We're not checking whether one assembly *gives access to another*. We're checking if one assembly *has access to another*. Those are indeed reversed tests; and this test is for the latter, not the former as you surmise.

Reply:
I will rename the parameters for clarity.

---
In reply to: [185161731](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161731) [](ancestors = 185161731)